core/src/main/java/org/apache/calcite/rel/rules/SubQueryRemoveRule.java

```diff
@@ -695,7 +715,7 @@ private static void matchJoin(SubQueryRemoveRule rule, RelOptRuleCall call) {
     Config JOIN = EMPTY
         .withOperandSupplier(b ->
             b.operand(Join.class)
-                .predicate(RexUtil.SubQueryFinder::containsSubQuery)
+                .predicate(SubQueryRemoveRule::hasSubQueryAndJoinTypeIsRewritable)
```

Comment:
Two other alternatives for the predicate; not sure if they are better than the current proposal I just like when methods do only one thing:
```
(join) -> isRewritable(join) && RexUtil.SubQueryFinder.containsSubQuery(join)
(join) -> !join.getJoinType().generatesNullsOnLeft && RexUtil.SubQueryFinder.containsSubQuery(join)
```
the second is not 100% equivalent but I am not sure if we intent to cover `CROSS`, `LEFT_SEMI_JOIN`, `COMMA`, if it ever reaches this rule.

Not necessarily requesting a change, leaving the final decision to @jamesstarr !

Reply:
The call is reused in assert in matchJoin.  So moving to the first one would mean, that it is repeated twice the `&& containsSubQuery` twice while still having a helper function.  Concerning `!join.getJoinType().generatesNullsOnLeft()`, the whole expression would have to be repeated twice, and while unlikely to change, it could allow for drift if some one wants to change the restrictiveness.  Furthermore, the phrasing is not particular intuitive.

I will rename hasSubQueryAndJoinTypeIsRewritable to containsSubQueryAndJoinTypeIsRewritable to match the existing verbs and change it use generatesNullsOnLeft() so it will likely work if any further join types are added.